File: frontend/src/composables/dataset.js
/*
*  Copyright (c) 2013-2025. The Johns Hopkins University Applied Physics Laboratory LLC
*
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* NO WARRANTY.   THIS MATERIAL IS PROVIDED "AS IS."  JHU/APL DISCLAIMS ALL
* WARRANTIES IN THE MATERIAL, WHETHER EXPRESS OR IMPLIED, INCLUDING (BUT NOT
* LIMITED TO) ANY AND ALL IMPLIED WARRANTIES OF PERFORMANCE,
* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT OF
* INTELLECTUAL PROPERTY RIGHTS. ANY USER OF THE MATERIAL ASSUMES THE ENTIRE
* RISK AND LIABILITY FOR USING THE MATERIAL.  IN NO EVENT SHALL JHU/APL BE
* LIABLE TO ANY USER OF THE MATERIAL FOR ANY ACTUAL, INDIRECT,
* CONSEQUENTIAL, SPECIAL OR OTHER DAMAGES ARISING FROM THE USE OF, OR
* INABILITY TO USE, THE MATERIAL, INCLUDING, BUT NOT LIMITED TO, ANY DAMAGES
* FOR LOST PROFITS.
 */

import {$api, getRequest} from "boot/axios";

export async function isActive(dataSetId) {
  const dataSet = (await getRequest( 'dataset/' + dataSetId)).data

  return dataSet.is_active
}
/**
 * Extract the dataset filters from the request logic.
 */
export function extractDataSetFilters(dataSet) {

  let filters = {}

  const request = dataSet.base_query.request
  if (request) {
    request.$and.forEach(fieldRequest => {
      // We need to pull the field name out of the field request. References have a different data structure than
      // standard and/or queries.
      let firstKey = Object.keys(fieldRequest)[0]
      let fieldName
      if (['$and', '$or'].includes(firstKey)) {
        const conditions = fieldRequest[firstKey]
        const firstCondition = conditions[0]
        fieldName = Object.keys(firstCondition)[0]
      }
      else {
        fieldName = firstKey
      }
      filters[fieldName] = fieldRequest
    })
  }
  return filters
}

/**
 * Create an easy way to access fields for a dataset
 */
export function createFieldsLookup(dataSet) {

  let fields = {}
  for (const field of dataSet.fields) {
    fields[field.data_field_name] = field
  }
  return fields
}

/**
 * Create an easy lookup for reference object values
 */
export function createReferenceOptionsLookup(dataSet) {

  let referenceOptions = {}
  const fields = dataSet.fields;

  fields.forEach(field =>  {
    if (field.region_map !== null && field.region_map !== undefined) {
      referenceOptions[field.data_field_name] = Object.keys(field.region_map_mapping).map((x) => ({label: x, value: x}))
    }
    else if (field.is_reference && field.values) {

      const options = field.values.map(function (fieldValue) {
        return {label: fieldValue === null ? '__null__' : fieldValue, value: fieldValue}
      })
      referenceOptions[field.data_field_name] = options;
    }
  })
  return referenceOptions;
}


/**
 * Go over the projection and get the shared fields.
 */
export function getSharedFields(dataSet) {
  let sharedFields = []

  dataSet.fields.forEach(field => sharedFields.push(field.data_field_name))

  return sharedFields
}

export async function query(dataSetBaseQuery, queryOpts, dataSetId) {
  let dataSetQuery = JSON.parse(JSON.stringify(dataSetBaseQuery))

  if (queryOpts.queryRowsPerPage) {
    dataSetQuery.limit = queryOpts.queryRowsPerPage
  }
  if (queryOpts.queryPage && queryOpts.queryRowsPerPage) {
    dataSetQuery.offset = queryOpts.queryRowsPerPage * (queryOpts.queryPage - 1)
  }
  if (queryOpts.querySortBy !== null) {
    dataSetQuery.order_by = [[queryOpts.querySortBy, queryOpts.queryDescending ? 'desc' : 'asc']]
  }

  if (queryOpts.queryGroupBy !== null
    && ((typeof queryOpts.queryGroupBy === 'string')
      || Array.isArray(queryOpts.queryGroupBy))) {

    dataSetQuery.group_by = {
      fields: Array.isArray(queryOpts.queryGroupBy) ? queryOpts.queryGroupBy : [queryOpts.queryGroupBy],
      aggregators: {
        [queryOpts.queryAggregateFunction]: {
          field: queryOpts.queryAggregateField,
          'function': queryOpts.queryAggregateFunction === 'rows' ? 'count' : queryOpts.queryAggregateFunction
        }
      }
    }
    if (queryOpts.aggregate){
      dataSetQuery.aggregate = queryOpts.aggregate
    }

    if (!await isActive(dataSetId)) {
      return {
        "data":[{
          "error": "inactive",
          "total": 0,
          "values": []
        }]
      }
    }

    const response = await $api.post( 'dataset/' + dataSetId + '/query', dataSetQuery)

    return response;
  }
}

File: backend/src/vims/app/model/dataset.py
#  Copyright (c) 2013-2025. The Johns Hopkins University Applied Physics Laboratory LLC
#
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# NO WARRANTY.   THIS MATERIAL IS PROVIDED "AS IS."  JHU/APL DISCLAIMS ALL
# WARRANTIES IN THE MATERIAL, WHETHER EXPRESS OR IMPLIED, INCLUDING (BUT NOT
# LIMITED TO) ANY AND ALL IMPLIED WARRANTIES OF PERFORMANCE,
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT OF
# INTELLECTUAL PROPERTY RIGHTS. ANY USER OF THE MATERIAL ASSUMES THE ENTIRE
# RISK AND LIABILITY FOR USING THE MATERIAL.  IN NO EVENT SHALL JHU/APL BE
# LIABLE TO ANY USER OF THE MATERIAL FOR ANY ACTUAL, INDIRECT,
# CONSEQUENTIAL, SPECIAL OR OTHER DAMAGES ARISING FROM THE USE OF, OR
# INABILITY TO USE, THE MATERIAL, INCLUDING, BUT NOT LIMITED TO, ANY DAMAGES
# FOR LOST PROFITS.

from typing import Any, Dict, List, Union

from datetime import datetime
from enum import Enum
from uuid import uuid4

from pydantic import Field

from vims.util import EnumStrLower

from .base import BaseModel


class DateGranularityEnum(EnumStrLower):
    DAILY = EnumStrLower.auto()
    WEEKLY = EnumStrLower.auto()
    MONTHLY = EnumStrLower.auto()
    YEARLY = EnumStrLower.auto()
    EPIWEEK = EnumStrLower.auto()


class DatasetField(BaseModel):
    display_name: str
    data_field_name: str
    data_field_type: str
    date_granularity: DateGranularityEnum | None = None
    is_reference: bool
    values: List[Union[str, None]] | None = None
    region_map_id: str | None = None
    region_map_mapping: Dict[str, str] | None = None

    class Config:
        use_enum_values = True


class DataList(BaseModel):
    value: List[float]


class OrderEnum(Enum):
    ascending = "asc"
    descending = "desc"


class DatasetQuery(BaseModel):
    computed: Dict[str, Dict[str, Union[str, List[str]]]] | None = None
    request: Dict[str, Any] | None = None
    projection: Dict[str, bool] | None = None
    group_by: Dict[str, Any] | None = None
    limit: int | None = None
    offset: int = 0
    order_by: List[List[Union[str, OrderEnum]]] | None = None
    transformations: Dict[str, Any] | None = None
    count_fields: List[str] | None = None
    distinct_field: str | None = None

    class Config:
        use_enum_values = True


class DatasetBase(BaseModel):
    dataset_name: str
    dataset_display_name: str
    description: str | None = None
    display_name: str
    # is_aggregate: bool
    datasource_id: str
    fields: List[DatasetField] = Field(default_factory=list)
    is_active: bool
    expiration: datetime | None = None
    primary_key_field: str | None = None
    date_field: str | None = None
    shared_with: List[str] | None = None
    groups: List[str] | None = None


class DatasetInternal(DatasetBase):
    last_modified: datetime | None = None
    last_modified_by_user: str


class Dataset(DatasetInternal):
    id: str = Field(default_factory=lambda: str(uuid4()))
    created: datetime = Field(default_factory=datetime.utcnow)
    created_by_user: str


class DatasetBaseQuery(DatasetBase):
    base_query: DatasetQuery | None = None


class DatasetAdminInternal(DatasetBaseQuery):
    id: str = Field(default_factory=lambda: str(uuid4()))
    last_modified: datetime | None = None
    last_modified_by_user: str


class DatasetAdmin(DatasetAdminInternal):
    created: datetime = Field(default_factory=datetime.utcnow)
    created_by_user: str


class DatasetRecord(DatasetBase):
    record: Dict[str, Any]


class DatasetRecordDelete(BaseModel):
    primary_key_field: str
    primary_key_value: Any
